app/assets/javascripts/discourse/app/initializers/live-development.js

```diff
@@ -60,12 +60,14 @@ export default {
             // Refresh if necessary
             document.location.reload(true);
           } else if (me.new_href && me.target) {
-            const link_target = me.theme_id
-              ? `[data-target="${me.target}"][data-theme-id="${me.theme_id}"]`
-              : `[data-target="${me.target}"]`;
-            document.querySelectorAll(`link${link_target}`).forEach((link) => {
-              this.refreshCSS(link, me.new_href);
-            });
+            const link_target = !!me.theme_id
+              ? `[data-target='${me.target}'][data-theme-id='${me.theme_id}']`
+              : `[data-target='${me.target}']`;
+
+            const links = document.querySelectorAll(`link${link_target}`);
```

Comment:
This is an interesting theory: perhaps we are getting issues because many extra stylesheets are somehow inserted into the DOM and this loop goes through them all. In that case we should try and figure out how that's happening.

I do have one suggestion though - rather than `querySelectorAll` you can use `querySelector` to find the first element, or `null` if it doesn't exist. I think it will mean less work for the browser.

Reply:
It's exactly that, many extra stylesheets are getting inserted into the DOM in some conditions and they are also causing the `forEach` loop to go haywire. So the fix in this PR is only partial, we have to track down what is inserting the many stylesheets in the DOM. (I am still investigating, working on a separate PR.)

I kept `querySelectorAll` here, because with `querySelector` we would be replacing the first stylesheet only. When testing with many quick subsequent updates to a theme that was causing problems. It's working better for me when the selector targets the last match, clones it with the new URL and removes it after a shorter timeout.

Reply:
we could querySelector and change it to search for :last or something? 

hmmm ... how is it possible to have the same link more than once on a page? Can we repro that?

Reply:
> querySelector and change it to search for :last or something?

I'm not quite sure how much browser support there is for a `[data-*]:last` or equivalent selector.

Yes, I can repro links multiplying on the page, that's what the change in https://github.com/discourse/discourse/pull/13755/commits/24395325216a5d50b6df4790b57c08119b867351, under `refreshCSS` is supposed to address. 

With a combination of many stylesheets updating at the same time plus multiple quick changes, the 2s timeout in `refreshCSS` can sometimes not clean up old `<link>` tags. On the next file change, the `forEach` loop would find the duplicates, try to replace all of them, and so on, exponentially.

Reply:
oh then maybe the big bug to fix here is to address the multiplication? if we ever see more than one link clean the old one up, log so we can fix the root cause?